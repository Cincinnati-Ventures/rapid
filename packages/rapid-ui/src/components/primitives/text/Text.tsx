import React from 'react';
import { RapidStyles } from '../../../utils';

type TextType =
	| 'p'
	| 'span'
	| 'i'
	| 'b'
	| 'u'
	| 'abbr'
	| 'cite'
	| 'kbd'
	| 'mark'
	| 's'
	| 'samp'
	| 'sup';

interface HeadingProps extends React.HTMLAttributes<HTMLParagraphElement> {
	styles?: string;
	as?: TextType;
}

const THEME_CLASSNAME = 'rapid-text';

const Text = React.forwardRef<HTMLParagraphElement, HeadingProps>(
	({ styles, as = 'p', ...rest }, ref) => {
		const Component = (type: TextType) => {
			switch (type) {
				case 'p':
					return (
						<p
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'span':
					return (
						<span
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'i':
					return (
						<i
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'b':
					return (
						<b
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'u':
					return (
						<u
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'abbr':
					return (
						<abbr
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'cite':
					return (
						<cite
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'kbd':
					return (
						<kbd
							ref={ref}
							{...rest}
							className={RapidStyles(styles, THEME_CLASSNAME)}
						/>
					);
				case 'mark':
					return (
						<mark
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 's':
					return (
						<s
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'samp':
					return (
						<samp
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
				case 'sup':
					return (
						<sup
							ref={ref}
							{...rest}
							className={RapidStyles(
								styles || rest.className,
								THEME_CLASSNAME,
							)}
						/>
					);
			}
		};
		return Component(as);
	},
);

Text.displayName = 'Text';

export default Text;
