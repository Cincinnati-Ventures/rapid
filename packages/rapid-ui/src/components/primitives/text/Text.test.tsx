import React from 'react';
import { render } from '@testing-library/react';
import { expect, it, describe } from 'vitest';
import Text from './Text';

describe('Text', () => {
	it('should render default correctly', () => {
		const { container } = render(<Text />);
		expect(container).toMatchSnapshot();
	});
});
