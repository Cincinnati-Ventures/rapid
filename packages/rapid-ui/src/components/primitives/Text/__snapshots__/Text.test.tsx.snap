// Vitest Snapshot v1

exports[`Text > should render default correctly 1`] = `
<div>
  <p
    class="rapid-text"
  />
</div>
`;
